### ISSUE TITLE
resilience-core OSGi bundle imports javax.annotation.meta

### ISSUE THREAD

**Issue by davsclaus** [resilience4j/resilience4j#750]
Thanks for raising a Resilience4j issue.
Please provide a brief description of your problem along with the versions you are using.
If possible, please also consider putting together a complete JUnit test that reproduces the issue.

Resilience4j version:
1.10.0

Java version:
8

Problem description:

In the MANIFEST.MF file you import
Import-Package: javax.annotation,javax.annotation.meta

Not sure what you use the .meta package for, as its not a standard package

In Apache Camel if we add a osgi feature for this then it fails validating
```
[ERROR] Failed to execute goal org.apache.karaf.tooling:karaf-maven-plugin:4.2.7:verify (validate) on project apache-camel: Verification failures: Verification failures:
[ERROR] 	Feature resolution failed for [camel-resilience4j/3.0.0.SNAPSHOT]
[ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=camel-resilience4j; type=karaf.feature; version=3.0.0.SNAPSHOT; filter:="(&(osgi.identity=camel-resilience4j)(type=karaf.feature)(version>=3.0.0.SNAPSHOT))" [caused by: Unable to resolve camel-resilience4j/3.0.0.SNAPSHOT: missing requirement [camel-resilience4j/3.0.0.SNAPSHOT] osgi.identity; osgi.identity=org.apache.camel.camel-resilience4j; type=osgi.bundle; version="[3.0.0.SNAPSHOT,3.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve org.apache.camel.camel-resilience4j/3.0.0.SNAPSHOT: missing requirement [org.apache.camel.camel-resilience4j/3.0.0.SNAPSHOT] osgi.wiring.package; filter:="(osgi.wiring.package=io.github.resilience4j.bulkhead)" [caused by: Unable to resolve io.github.resilience4j.bulkhead/1.1.0: missing requirement [io.github.resilience4j.bulkhead/1.1.0] osgi.wiring.package; filter:="(osgi.wiring.package=io.github.resilience4j.core.exception)" [caused by: Unable to resolve io.github.resilience4j.core/1.1.0: missing requirement [io.github.resilience4j.core/1.1.0] osgi.wiring.package; filter:="(osgi.wiring.package=javax.annotation.meta)"]]]]
[ERROR] Repositories: {
[ERROR] 	file:///Users/davsclaus/workspace/camel/platforms/karaf/features/target/classes/features.xml
```

**Comment by RobWin** [comment]
Thanks,

yes, it seems we have to configure the Gradle OSGi plugin properly.

**Comment by davsclaus** [comment]
Some of my OSGi friends told me that

```
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
```

means Java 1.8 only. So if you intend to support Java 9, 10, 11 etc then maybe better to remove that line too.

(yeah osgi manifest files are pain)

**Comment by davsclaus** [comment]
Ah sorry now I was told that this actually means 1.8 or better on Apache Karaf. But there are some ambiguity in OSGi. Normally you need version ranges, eg [1.8,) to say 1.8 or better. And then Java version are changed to be, 1.9, 9.0, 10.0, 11.0 and so on. So its only the top problem with that meta annotation import.

**Comment by davsclaus** [comment]
That meta annotation seems to come from jsr-305, which seems to be used by findbugs or some defect reporting tools, so maybe this helps track down the culprit.

**Comment by RobWin** [comment]
Yes, thanks. The problem is that the plugin isn't configured. We are using defaults, which we should change now.

**Comment by Freedooom111** [comment]
To fx89 t christogram 9f7f6

**Comment by cziegeler** [comment]
The problem with the annotations is that they have a retention of runtime, so they are required at runtime in order to load the classes properly.
Some annotations in the io.github.resilience4j.core.lang package use annotations from javax.annotation.meta .
The annotations in io.github.resilience4j.core.lang have retention runtime as well, so they are needed at runtime - assuming that other libraries use these annotations.
The obvious fix would be to remove usage of javax.annotation.meta - not sure for what the annotations are used though

**Comment by myroch** [comment]
Upgrade to Java 9 will be probably also more complex, just because of these annotations. This issue is opened almost one year now, can we expect some progress on it soon?
JSR-305 Annotations are outdated, resp. never accepted anyway - see https://blog.codefx.org/java/jsr-305-java-9/ and following comments.

**Comment by jettro** [comment]
We are running into this issue at the moment as well. We cannot really include the com.google.code.findbugs:jsr305:3.0.2 which seems to be required in an OSGi container. The problem is that this jar includes the javax.annotation-api.Resource which is also included in the javax.annontation-api.jar. So we get OOM errors when running in felix. Not an easy way around it. Would it be possible to remove this runtime dependency somehow?

**Comment by Romeh** [comment]
I will have a deeper look on that this week to see what what we can do

**Comment by alienintheheights** [comment]
Same issue. Trying to use this in an Apache Karaf OSGi module.

**Comment by RobWin** [comment]
Hi, I think I need some support. Someone interested to fix it and create a PR?

**Comment by ehimsi** [comment]
Hi All 
A workaround  for OSGi based platforms - AEM and Karaf. If you are using maven-bundle-plugin, then use this in the pom - 
`<plugin>
                <groupId>org.apache.felix</groupId>
                <artifactId>maven-bundle-plugin</artifactId>
                <extensions>true</extensions>
                <configuration>
                    <instructions>
                        <Embed-Directory>target/dependency</Embed-Directory>
                        <Import-Package>
                            
                            !kotlin,
                            !javax.annotation.meta,
                            !javax.enterprise.util,`

**Comment by cziegeler** [comment]
I've created a potential PR #1402 
It changes the retention policy of the annotations to CLASS, so they are not needed at runtime anymore.
And it does not export the annotation package (core.lang) from the core bundle anymore.
I think this could work, but for sure needs more eyes for review

**Comment by RobWin** [comment]
Thx. Did you try to verify it locally? 
You could build the jar locally and check the generated MANIFEST.
But I guess you know it better then me ;)

**Comment by cziegeler** [comment]
The manifest is "correct" and I can install the bundle and it resolves. Now, the tricky part is that with not exporting core.lang anymore other existing bundles currently importing that package do not resolve anymore. For example, if you use the new core bundle with an old resilience4j cache bundle, the cache bundle will not resolve. So it's an all or nothing approach, you would need to also use the new version of the cache bundle

**Comment by alienintheheights** [comment]
I was able to export my bundle with the following. Mostly needed to exclude these packages from the export. I also import everything as optional. Other bundles in our karaf environment are now able to use this one without any knowledge of resilience or vavr dependencies. Not an expert on this so if there's a better way, please let me know. 

```
       <plugin>
              <groupId>org.apache.felix</groupId>
                <artifactId>maven-bundle-plugin</artifactId>
                <extensions>true</extensions>
                <configuration>
                    <instructions>
                      <Embed-Dependency>,
                          *;scope=compile|runtime</Embed-Dependency>
                      <Embed-Transitive>true</Embed-Transitive>
                      <Export-Package>
                      	  my.package.com,
                          !io.github.resilience4j.*,
                          !io.vavr.*
                      </Export-Package>
                      <Import-Package>
                        *;resolution:=optional
                      </Import-Package>
                    </instructions>
                </configuration>
             </plugin>
```

**Comment by cziegeler** [comment]
Any feedback on the PR?
I think the PR is a correct solution, it has the following changes:
- change the retention policy to CLASS for the annotations in  resilience4j-core/src/main/java/io/github/resilience4j/core/lang - this should do no harm as these annotations are used by IDE and build tools
- resilience-core : do not export io.github.resilience4j.core.lang - again no harm as these are not needed at runtime
- resilience-core : avoid importing javax.annotation and javax.annotation.meta - these packages are not available at runtime anyway
- all other bundles : avoid importing   javax.annotation and javax.annotation.meta, and io.github.resilience4j.core.lang - again, these are not used at runtime 

The only minor disadvantage is that you can't use a new core bundle with an older other bundle like cache which imports io.github.resilience4j.core.lang - you need to use the new version of cache as well., But as this hasn't really worked well with old versions in OSGi anyway, I don't think it's  a problem.

**Comment by alienintheheights** [comment]
Related to this, I was able to get a bundle that includes the resilience4j jar working with other bundles in a Karaf container via the following Felix block in its pom.xml. I'm not an expert on OSGi so perhaps I've been a little too wide on the import side. But at least the other bundles are able to consume this without knowledge of these jars now. 

```
    <plugin>
        <groupId>org.apache.felix</groupId>
        <artifactId>maven-bundle-plugin</artifactId>
        <extensions>true</extensions>
        <configuration>
           <instructions>
           <Embed-Dependency>
                          *;scope=compile|runtime</Embed-Dependency>
           <Embed-Transitive>true</Embed-Transitive>
           <Export-Package>
                      	   com.mypackage,
                          !io.github.resilience4j.*,
                          !io.vavr.*
            </Export-Package>
            <Import-Package>
                        *;resolution:=optional
           </Import-Package>
         </instructions>
     </configuration>
 </plugin>
```
